### PR TITLE
Fix one memory leak issue

### PR DESCRIPTION
**Pull request by qwu16**

### CODE REVIEW

File: source/agent/analytics/videoGstPipeline/VideoGstAnalyzer.cc [hunk 1]
```diff
@@ -76,7 +75,7 @@ gboolean VideoGstAnalyzer::StreamEventCallBack(GstBus *bus, GstMessage *message,
             ELOG_DEBUG("State change from %d to %d, play:%d \n",old_state, new_state, GST_STATE_PAUSED);
             if(new_state == GST_STATE_PLAYING || new_state == GST_STATE_PAUSED) {
                 ELOG_DEBUG("Create debug dot file\n");
-                GST_DEBUG_BIN_TO_DOT_FILE_WITH_TS(GST_BIN(pStreamObj->pipeline),GST_DEBUG_GRAPH_SHOW_ALL, "play");
+                //GST_DEBUG_BIN_TO_DOT_FILE_WITH_TS(GST_BIN(pStreamObj->pipeline),GST_DEBUG_GRAPH_SHOW_ALL, "play");
```
Comment (taste1981) [inline source/agent/analytics/videoGstPipeline/VideoGstAnalyzer.cc]: for code not used, please remove direclty.

File: source/agent/analytics/videoGstPipeline/VideoGstAnalyzer.cc [hunk 2]
```diff
@@ -98,7 +97,7 @@ void VideoGstAnalyzer::clearPipeline()
  
     }
 
-int VideoGstAnalyzer::createPipeline() {
+bool VideoGstAnalyzer::createPipeline() {
 
     loop = g_main_loop_new(NULL, FALSE);
```
Comment (taste1981) [inline source/agent/analytics/videoGstPipeline/VideoGstAnalyzer.cc]: Please create the g_main_loop after pipeline_ is successfully created. You unref conditionally in clearPlugin, so there might be a memory leak here.

File: source/agent/analytics/analytics-agent.js [hunk 3]
```diff
@@ -105,7 +105,9 @@ createInternalConnection(connectionId, direction, internalOpt) {
                keyFrameInterval, 'bitrate:',bitrate);
       
       this.engine.setOutputParam(codec,resolution,framerate,bitrate,keyFrameInterval,algo,pluginName);
-      this.engine.createPipeline();
+      if(this.engine.createPipeline() === false) {
```
Comment (taste1981) [inline source/agent/analytics/analytics-agent.js]: spacing.

File: source/agent/analytics/videoGstPipeline/VideoGstAnalyzer.cc [hunk 4]
```diff
@@ -224,14 +180,21 @@ void VideoGstAnalyzer::new_sample_from_sink (GstElement * source, gpointer data)
...
     gst_buffer_unmap(buffer, &map);
+    gst_sample_unref(sample);
 }
 
 int VideoGstAnalyzer::addElementMany() {
-    if(pipeline_)
-        pipeline_->LinkElements();
+    if(pipeline_){
+        rvaStatus status;
+        status = pipeline_->LinkElements();
```
Comment (taste1981) [inline source/agent/analytics/videoGstPipeline/VideoGstAnalyzer.cc]: combine to rvaStatus status = pipeline_LinkEelements();

File: source/agent/plugins/samples/cpu_pipeline/mypipeline.cc [hunk 5]
```diff
@@ -118,8 +118,8 @@ rvaStatus MyPipeline::LinkElements() {
 
     GstCaps* encodecaps = gst_caps_new_simple("video/x-h264",
 		"stream-format", G_TYPE_STRING, "byte-stream",
-		"width", G_TYPE_INT, 640,
-		"height", G_TYPE_INT, 480,
+		"width", G_TYPE_INT, inputwidth,
```
Comment (taste1981) [inline source/agent/plugins/samples/cpu_pipeline/mypipeline.cc]: please follow the same style for the variable names.

File: source/core/owt_base/InternalOut.cpp [hunk 6]
```diff
@@ -33,15 +33,15 @@ void InternalOut::onFrame(const Frame& frame)
...
     memset(&outFrame, 0, sizeof(outFrame));
 
     outFrame.format = FRAME_FORMAT_H264;
     outFrame.length = length;
-    outFrame.additionalInfo.video.width = 1440;
-    outFrame.additionalInfo.video.height = 810;
+    outFrame.additionalInfo.video.width = width;
+    outFrame.additionalInfo.video.height = height;
     if(m_frameCount == 0)
        outFrame.additionalInfo.video.isKeyFrame = true;
```
Comment (taste1981) [inline source/core/owt_base/InternalOut.cpp]: why is that?

File: source/core/owt_base/InternalOut.cpp [hunk 7]
```diff
@@ -33,15 +33,15 @@ void InternalOut::onFrame(const Frame& frame)
     m_transport->sendData(sendBuffer, header_len + 1, reinterpret_cast<char*>(const_cast<uint8_t*>(frame.payload)), frame.length);
 }
 
-void InternalOut::onFrame(uint8_t *buffer, uint32_t length)
+void InternalOut::onFrame(uint8_t *buffer, int width, int height, uint32_t length)
```
Comment (starwarfan) [inline source/core/owt_base/InternalOut.cpp]: The internalOut already has a "onFrame(const Frame&)" function, why you add a function and construct frame struct inside?
Why not construct a frame struct directly and call the previouse "onFrame" function?
Like "x->onFrame(outFrame)"

File: source/core/owt_base/InternalOut.h [hunk 8]
```diff
@@ -19,7 +19,7 @@ class InternalOut : public FrameDestination, public RawTransportListener {
     virtual ~InternalOut();
 
     void onFrame(const Frame&);
-    void onFrame(uint8_t *buffer, uint32_t length);
+    void onFrame(uint8_t *buffer, int width, int height, uint32_t length);
```
Comment (starwarfan) [inline source/core/owt_base/InternalOut.h]: This is  unnecessary

File: source/agent/analytics/videoGstPipeline/VideoGstAnalyzer.cc [hunk 9]
```diff
@@ -14,7 +14,6 @@ namespace mcu {
 DEFINE_LOGGER(VideoGstAnalyzer, "mcu.VideoGstAnalyzer");
 
 GMainLoop* VideoGstAnalyzer::loop = g_main_loop_new(NULL,FALSE);
```
Comment (starwarfan) [inline source/agent/analytics/videoGstPipeline/VideoGstAnalyzer.cc]: You initialize the static variable loop here, then why you do it again at line 134?

File: source/agent/analytics/videoGstPipeline/VideoGstAnalyzer.cc [hunk 10]
```diff
@@ -135,10 +127,11 @@ int VideoGstAnalyzer::createPipeline() {
     pipeline = pipeline_->InitializePipeline();
 
     if (!pipeline) {
-        ELOG_ERROR("pipeline could not be created\n");
+        ELOG_ERROR("pipeline Initialization failed\n");
         return -1;
     }
 
+    loop = g_main_loop_new(NULL, FALSE);
```
Comment (starwarfan) [inline source/agent/analytics/videoGstPipeline/VideoGstAnalyzer.cc]: The loop is a static memeber variable? Is this duplicated as line 16?

**Pull request merged**